Add -Raw switch to Select-String for convenience

Jawz84 opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

Resolves #7713 
This adds a `-Raw` switch parameter to Select-String, for convenience. 

<!-- Summarize your PR between here and the checklist. -->

## PR Context

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->
This is my first PR to the PowerShell repo, so I picked this tiny issue. As to why this switch is desirable has already been discussed thoroughly in the issue. 

~~The approach I've taken does not introduce new parameter sets, even though -Raw is not compatible with -Quiet and -Context.~~ In the case -Raw is used together with -Context, the latter is ignored. To me, it would not make sense to have context lines when all we want is matched lines. ~~In the case -Raw is used together with -Quiet, Quiet takes precedence, because the choice for precedence seemed somewhat arbitrary and it was easier to implement this way. One might argue that respecting a users whish for -Quiet behaviour is more important than returning strings in that case.~~
_Edit: after discussion, -Raw and -Quiet were put in separate parametersets._

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.

I'm unsure about this part below ⬇️ . I think you could say this is a backwards compatible breaking change. It introduces a new feature, but does not break existing functionality. Please advise.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/4537
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

Commit:
Add -Raw switch for convenience (returns only matched lines as string)

Commit:
Add tests for -Raw switch

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=9901) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=9901) before we can accept your contribution.<br/><br/>:x: Jawz84 [sign now](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=9901)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/PowerShell/PowerShell?pullRequest=9901) it.</sub>

vexx32 commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -1019,6 +1019,13 @@ public string[] LiteralPath
+        /// If not (default) return MatchInfo (or bool objects, when Quiet is passed).
+        /// </summary>
+        [Parameter]
+        public SwitchParameter Raw { get; set; }
```
This should probably be in a separate parameter set to the `Quiet` switch, otherwise behaviour when both are supplied will be confusing regardless of which is provided.

They both transform the output, so we shouldn't allow them to be present at the same time imo. This would also let you simplify your logic further down a bit by checking `ParameterSetName` before looking at Raw/Quiet

Jawz84 replied:
Thanks for spotting that, I've thought about this. Did you see my comment here: 

> The approach I've taken does not introduce new parameter sets, even though -Raw is not compatible with -Quiet and -Context. In the case -Raw is used together with -Context, the latter is ignored. To me, it would not make sense to have context lines when all we want is matched lines. In the case -Raw is used together with -Quiet, Quiet takes precedence, because the choice for precedence seemed somewhat arbitrary and it was easier to implement this way. One might argue that respecting a users whish for -Quiet behaviour is more important than returning strings in that case.

The -List and -Quiet params did not go in separate parameter sets either. That also made me swing this way.

vexx32 replied:
I hear you, and maybe the PS team will have a different opinion, but in my mind it's a lot better from a UX standpoint to have these separated out in different sets. 

It's too easy to get confused as to which output takes precedence. We have the ability to separate them out, so I'd say... why not? 😄

Jawz84 replied:
OK, no worries. I'll put -Raw and -Quiet in different paramsets then 😉 . First thing tomorrow morning!

SteveL-MSFT replied:
I would just say that don't assume the current code is the best way to do something :)  PowerShell is a large code base written by a large number of people over a large number of years.  It's generally rare that old code is updated when new and better patterns emerge.  In this specific case, my preference is to have different parameter sets and consider fixing `-List` and `-Quiet` (unless there's a clear reason why it was done this way, but doesn't have to be part of this PR, can be separate one).  Note that we do have specific cases where we decided it was best to not have parametersets and do a runtime check due to the complexity of many parameters that were mutually exclusive over several existing parametersets, but those should be exceptions.

vexx32 replied:
@SteveL-MSFT I think `Where-Object` begs to differ. I've seen those very very naughty parameter sets 😉

Jawz84 replied:
param sets it is!

Jawz84 replied:
@SteveL-MSFT I've implemented the requested changes. CI is happy now. I'm not sure how to mark your requested change as 'done' :)

Jawz84 commented:
I'll need some guidance as to what I am supposed to do to fix any failing CI requirements. Please let me know what I can do as a next step. Thank you 🙂.

vexx32 reviewed (commented):
Love it! One thing to note but other than that this looks good. :smile:

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1:
```diff
@@ -1,6 +1,6 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
-Describe "Select-String" -Tags "CI" {
+Describe "Select-String" -Tags "CI","SLS" {
```
Why is there a `SLS` tag?  We're using specific tags that our CI understands: https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/WritingPesterTests.md#describecontextit

Jawz84 replied:
I'll remove that - was there to shorten my test cycles. Has slipped in by accident.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1:
```diff
@@ -78,6 +78,11 @@ Describe "Select-String" -Tags "CI" {
 	    $equal | Should -Be True
 	}
+
+	it "Should return a string type when -Raw is used" {
```
```suggestion
	It "Should return a string type when -Raw is used" {
```

Jawz84 replied:
fixed.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1:
```diff
@@ -78,6 +78,11 @@ Describe "Select-String" -Tags "CI" {
+
+	it "Should return a string type when -Raw is used" {
+		$result = $testinputtwo | Select-String -Pattern "hello" -CaseSensitive -Raw
+		,$result | Should -BeOfType "System.String"
```
Why are you explicitly changing the left side to an array?

Jawz84 replied:
Tbh: because the other tests that were already there were set up like that too, and I did not want to break style. I'll remove it. It does not matter for the outcome of the test though.

Jawz84 committed:
Remove temporary SLS test tag

Co-Authored-By: Steve Lee <slee@microsoft.com>

Jawz84 commented:
I'll have to finish reading rest of the comments later (I have an appointment tonight). Thank you for getting back to me so fast! Much appreciated!

Commit:
Fix requested changes: capitalization, undo force array

Commit:
Put -Raw and -Quiet in separate parametersets, adjust tests.

Jawz84 commented:
Parametersets and accompanying test fixed. Just let me know if there is anything else that needs to be done 😉

daxian-dbw commented:
Close and reopen the PR to force Windows CI build working properly.

Jawz84 commented:
Hi Dongbo @daxian-dbw, I didn't know the trick with close/reopen issue to get CI to rerun. Would you like me to do that again? This time, the Mac OS test lost connection with the agent.

Jawz84 commented:
I'll close and reopen this once, to re-trigger CI, this clearly looks like a glitch. If that doesn't turn out all- green, I leave it to the team.

daxian-dbw commented:
@PoshChan please get failures for macOS

PoshChan commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -295,8 +295,8 @@ internal MatchInfo Clone()
     /// <summary>
```
@daxian-dbw, test results were not published for PowerShell-CI-linux at vstfs:///Build/Build/25352

PoshChan replied:
@daxian-dbw, test results were not published for PowerShell-CI-static-analysis at vstfs:///Build/Build/25389

PoshChan replied:
@daxian-dbw, test results were not published for PowerShell-CI-windows at vstfs:///Build/Build/25392

PoshChan commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -295,8 +295,8 @@ internal MatchInfo Clone()
     /// <summary>
```
@daxian-dbw, your last commit had 1 failures in `PowerShell-CI-macos`
Read-Host.[Console]::ReadKey() implementation on non-Windows.Should output correctly
```powershell
Expected 0, but got 1.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Host/Read-Host.Tests.ps1: line 25
25:             $LASTEXITCODE | Should -Be 0
```

daxian-dbw replied:
@PoshChan Please rerun for macOS

Jawz84 commented:
Mac OS failed again, where it passed the other day. Awaiting the PSTeam.

Jawz84 commented:
This PR seems blocked on CI glitches. Could you please have a look @anmenaga? If there is anything I can do, just let me know :). Thank you!

Jawz84 commented:
@PoshChan Please rerun for macOS

PoshChan commented:
@Jawz84, you are not authorized to request a rebuild

Jawz84 commented:
The requested changes have been implemented. CI is all green. @stevel_msft can you have another look? I think this is ready.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -1433,8 +1448,15 @@ private bool FlushTrackerQueue(IContextTracker contextTracker)
+                    WriteObject(match.Line);
+                }
+            }
             // If -quiet is specified but not -list return true on first match
```
Please remove the obvious comment.
```suggestion
```

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1:
```diff
@@ -78,6 +78,15 @@ Describe "Select-String" -Tags "CI" {
 	    $equal = @(Compare-Object $firstMatch $secondMatch).Length -eq 0
 	    $equal | Should -Be True
```
```suggestion
	    $equal | Should -BeTrue
```

Jawz84 committed:
Remove unnecessary comment.

Co-Authored-By: Ilya <darpa@yandex.ru>

Jawz84 committed:
Fix bug in existing Pester test

Co-Authored-By: Ilya <darpa@yandex.ru>

Jawz84 commented:
Thank you for your review @iSazonov.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1:
```diff
@@ -175,6 +184,24 @@ Describe "Select-String" -Tags "CI" {
+	    $expected3 = "This is the third line"
+	    $expected4 = "This is the fourth line"
+
+	    (Select-String in $testInputFile -Raw)[0] | Should -Be $expected1
```
Please use BeExactly for strings. Below too.
```suggestion
	    (Select-String in $testInputFile -Raw)[0] | Should -BeExactly $expected1
```

Jawz84 replied:
I don't think that is the right thing to do in this case: Select-String is case insensitive, unless `-CaseSensitive` is specified. Are you sure?

iSazonov replied:
We do not check "case insensitive". The check should guarantee that we received _exactly_ the original string. Perhaps it is not critical for the test but it is best practice.

Jawz84 replied:
OK. I see your point, and I agree now. Would you like me to retrofit this same fix to lines 133-137?

iSazonov replied:
You can do this in the PR (or move to follow PR).

Jawz84 committed:
Fix test to adhere to Pester best practice 'use -BeExactly for strings'.

Co-Authored-By: Ilya <darpa@yandex.ru>

Commit:
Retrofit -BeExactly instead of -Be for string tests

iSazonov commented:
@SteveL-MSFT Please update your review.

anmenaga commented:
@Jawz84 :
* Please open a [documentation issue](https://github.com/powershell/powershell-docs/) to document new parameter and updates to parameter sets; Please add a link to that doc issue to the description of this PR under "Issue filed:".
* The CodeFactor issue in the new code needs to be fixed - `Insert parentheses within the conditional AND and OR expressions to declare the operator precedence.` - line 1186 - `private IContextTracker GetContextTracker() => (Raw || _preContext == 0 && _postContext == 0) ? _noCon..`

@vexx32 Can you please look through this PR again and approve if you don't see anything blocking? Thank you.

vexx32 commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -972,7 +972,8 @@ void IContextTracker.TrackEOF()
         /// Gets or sets the current pipeline object.
         /// </summary>
-        [Parameter(ValueFromPipeline = true, Mandatory = true, ParameterSetName = "Object")]
+        [Parameter(ValueFromPipeline = true, Mandatory = true, ParameterSetName = "ObjectQuiet")]
```
Why are these sets named `ObjectQuiet` and `FileQuiet`? Isn't the normal behaviour with these parameter sets to output the MatchInfo objects? 

It should only need to be named after the `-Quiet` switch if that switch is the defining mandatory parameter for that set -- in this case it's just an optional switch for any of these sets.

I'd recommend just using `Object` / `File` / `ObjectRaw` / `FileRaw` names, as the `Quiet` appears to imply the default behaviour is `-Quiet`, which would probably not be desirable for future maintainability. 🙂

Jawz84 replied:
I'll have to look into this one later, thank you.

Jawz84 replied:
Yes, absolutely. After a good night's sleep, this makes lots of sense now. I've removed the 'Quiet' part of the parameternames, as you suggested. Good catch.

vexx32 reviewed (commented):
Apart from the parameter set names I think this is all good to go.

Nice work! 😄

Commit:
Add parens to make operator precedence explicit, fixing CodeFactor issue

Jawz84 commented:
@anmenaga Thank you, I've added parens to fix the CodeFactor issue. I've also put in a rudimentary issue in for documentation, and added it to the description above.

Commit:
Remove 'quiet' from parametersetnames, because it was misleading

Jawz84 commented:
Closing/reopening this to trigger CI. Mac os test ran into an unrelated error: 
```text
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Core/Job.Tests.ps1: line 108
108:                         $results = Receive-Job -Keep $job
```
https://powershell.visualstudio.com/PowerShell/_build/results?buildId=27521&view=ms.vss-test-web.build-test-results-tab&runId=1142426&resultId=104912&paneView=debug

vexx32 commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -296,7 +296,7 @@ internal MatchInfo Clone()
     /// </summary>
     [Cmdlet(VerbsCommon.Select, "String", DefaultParameterSetName = "File", HelpUri = "https://go.microsoft.com/fwlink/?LinkID=113388")]
-    [OutputType(typeof(MatchInfo), typeof(bool))]
+    [OutputType(typeof(MatchInfo), typeof(bool), typeof(string))]
```
If you look at `TestConnectionCommand` for a moment, it has a great example of how to work with OutputType -- you can actually specify the output type(s) based on the parameter set name.

I would redefine it to have the MatchInfo/bool type in Object & File sets, and then give it string output type for the ObjectRaw / FileRaw sets. :)

Jawz84 replied:
Oh, I did not know that was even possible! Neat! Will look at that, thanks.

anmenaga replied:
So looks like this is the only remaining thing for this PR and then we're merging it.  or let me know if this is not blocking.

Jawz84 replied:
I'm going to work on this tonight. I have no idea if this would be blocking or not. That's not up to me either I think 🙂

Jawz84 replied:
@vexx32, I've pushed changes, could you have a look please? The problem I ran into is, that because `-Quiet` is not in it's own paramter set, I don't think we can get `OutputType` to be more specific than this. What do you think?

Jawz84 replied:
I've also gone ahead and refactored the parameter set names to string constants, just like in TestConnectionCommand.cs. It makes the code more readable and easier to refactor later.

vexx32 reviewed (approved):
One minor nit, but otherwise this looks awesome. Great work! 😊

Commit:
Refactor OutputType to be more specific for given parameter set name

Also refactor parameter set names to string constants.

Commit:
Remove spaces and unnessesary type name to fix CodeFactor and Codacy issues

Jawz84 commented:
Close/Reopen issue to trigger CI. CI encountered a transient error, `Test-Connection` timeout:
`Expected the actual value to be less than 3, but got 3.0422223.`

vexx32 commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -295,10 +295,18 @@ internal MatchInfo Clone()
-    [OutputType(typeof(MatchInfo), typeof(bool))]
+    [Cmdlet(VerbsCommon.Select, "String", DefaultParameterSetName = ParameterSetFile, HelpUri = "https://go.microsoft.com/fwlink/?LinkID=113388")]
+    [OutputType(typeof(bool), typeof(MatchInfo), ParameterSetName = new[] { ParameterSetFile, ParameterSetObject, ParameterSetLiteralFile })]
+    [OutputType(typeof(bool), typeof(string), ParameterSetName = new[] { ParameterSetFileRaw, ParameterSetObjectRaw, ParameterSetLiteralFileRaw })]
```
`-Raw` and `-Quiet` can't be used together, right? I'd think this can just be `typeof(string)` as you shouldn't run into bool here.

Jawz84 replied:
Huh I left that in? Thanks, fixed.

vexx32 commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -1168,7 +1190,8 @@ public new int[] Context
 
-        private IContextTracker GetContextTracker() => (_preContext == 0 && _postContext == 0) ? _noContextTracker : new ContextTracker(_preContext, _postContext);
+        // When we are in Raw mode or pre- and postcontext are zero, use the _noContextTracker, since we will not be needing trackedLines.
+        private IContextTracker GetContextTracker() => (Raw || (_preContext == 0 && _postContext == 0)) ? _noContextTracker : new ContextTracker(_preContext, _postContext);
```
Since we're adding something here anyway, I feel it might be time to tidy this long line up a bit?

```suggestion
        private IContextTracker GetContextTracker() => (Raw || (_preContext == 0 && _postContext == 0)) ? _noContextTracker : new ContextTracker(_preContext, _postContext);
```
```suggestion
        private IContextTracker GetContextTracker() => (Raw || (_preContext == 0 && _postContext == 0)) 
            ? _noContextTracker
            : new ContextTracker(_preContext, _postContext);
```

Jawz84 replied:
Agreed, fixed.

vexx32 reviewed (commented):
Definitely looking much better! 🙂

Commit:
Add linebreaks to improve readability

Commit:
Remove typeof(bool) from OutputType for Raw parameter sets

Jawz84 commented:
@anmenaga over to you 🙂

anmenaga commented:
@Jawz84 Thank you for your contribution!
